src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java

```diff
@@ -21,22 +32,94 @@ public boolean canCheck(CaseDocument caseDocument) {
         return true;
     }
 
+    private boolean isDuplicateFilename(CaseDocument caseDocument, Supplier<List<DocumentFileNameProvider>> caseDocumentSupplier) {
+        return ofNullable(caseDocumentSupplier.get()).orElse(List.of(CaseDocument.builder().documentFilename("").build()))
+            .stream().anyMatch(d -> d.getDocumentFilename().equals(caseDocument.getDocumentFilename()));
+    }
+
+    private static void processList(List<?> list, List<DocumentFileNameProvider> allDocuments) {
+        if (list != null) {
+            for (Object item : list) {
+                if (item instanceof HasCaseDocument hasCaseDocument) {
+                    processHasCaseDocument(hasCaseDocument, allDocuments);
+                } else {
+                    log.warn("Ignored " + item.getClass().getName());
+                }
+            }
+        }
+    }
+
+    private static void processHasCaseDocument(HasCaseDocument hcd, List<DocumentFileNameProvider> allDocuments) {
+        if (hcd != null) {
+            try {
+                // Collect all fields from HasCaseDocument class
+                Field[] fields = hcd.getClass().getDeclaredFields();
+
+                for (Field field : fields) {
+                    field.setAccessible(true);
+
+                    // Check if the field is a List with a parameterized type
+                    if (List.class.isAssignableFrom(field.getType())) {
+                        ParameterizedType parameterizedType = (ParameterizedType) field.getGenericType();
+                        Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
+
+                        // Ensure the list has a single parameterized type argument
+                        if (actualTypeArguments.length == 1 && HasCaseDocument.class.isAssignableFrom((Class<?>) actualTypeArguments[0])) {
+                            // Get the value of the field and process the list
+                            processList((List<?>) field.get(hcd), allDocuments);
+                        }
+                    } else if (DocumentFileNameProvider.class.isAssignableFrom(field.getType())) {
+                        allDocuments.add((DocumentFileNameProvider) field.get(hcd));
+                    } else if (HasCaseDocument.class.isAssignableFrom(field.getType())) {
+                        processHasCaseDocument((HasCaseDocument)  field.get(hcd), allDocuments);
+                    }
+                }
+            } catch (Exception e) {
+                log.error("Failed to check for duplicate filenames and return warnings", e);
+            }
+        }
+    }
+
+    private List<DocumentFileNameProvider> collectCaseDocumentsFromFinremCaseData(FinremCaseData caseData) {
+        List<DocumentFileNameProvider> allDocuments = new ArrayList<>();
+        processHasCaseDocument(caseData, allDocuments);
+        return allDocuments.stream().filter(Objects::nonNull).toList();
+    }
+
```

Comment:
Can ln 53 be removed?
```
 if (hcd != null) {
```
Line 43 is doing the null check already as far as I can tell:

```
if (item instanceof HasCaseDocument hasCaseDocument) {
```

Reply:
Also, just to check, will the logs on line 46 and 78 be useful?  As in can we tie them back to a case if we were looking in Insights?

Reply:
Good spot. Yes, NPE has been handled and the log does not make any sense. I've just removed the check null logic and the log.